playbooks/development-environments.md

```diff
@@ -0,0 +1,96 @@
+- Projects default to configuration values that are appropriate for development (just as they run in "development"
+  mode by default).
+- Both local and containerized development are supported, with minimal duplication.
+- Sensitive values are never committed to repositories, whether open- or closed-source.
+- Other things being equal, less configuration, simpler-setup, and fewer dependencies are better than more.
+
+## Prerequisites
+
+Setup tooling is optimized for a default set of tools:
+
+- Mac computers
+- `bash` shell
+- [Homebrew](https://brew.sh/) for installing and managing supporting services
+- Docker for containerization
+- [nvm](https://github.com/nvm-sh/nvm) for managing node versions
+- [rbenv](https://github.com/rbenv/rbenv) for managing ruby versions
+- [hokusai](hokusai.md) for managing deployed applications
+- [VSCode](https://code.visualstudio.com/) for text editing
+- [awscli](https://aws.amazon.com/cli/) for AWS operations
+
+Developers who find other tools compelling may need to adapt setup procedures to their choices. For the sake of
+simplicity, we avoid making setup tooling heavily conditional on local choices.
+
+The [shared setup script](https://github.com/artsy/potential/blob/master/scripts/setup)🔒 installs these tools, and
+individual projects' setups may depend on them.
+
+These choices can and should be improved (by [RFC-style](rfcs.md) pull request), but care should be taken to update
+any dependent projects.
+
+## Project set-up
+
+Project's READMEs should clearly document the commands necessary for fresh set-up, ideally just reviewing and
+running a `./bin/setup` script (or whatever is idiomatic for that stack). The script should:
+
+- install dependencies such as data stores
+- install necessary language runtimes/versions
+- initialize configuration with values that are appropriate for development (ideally none)
+- install packages (rubygems, node modules, etc.)
+- create and seed databases, if appropriate
+- ideally be idempotent (i.e., re-running setup should work to update an existing environment)
+- be clearly organized and commented, so it can serve as a guide in incompatible environments
+
+## Shared configuration
+
+We strive to avoid committing any sensitive values to repositories, but sometimes they are necessary for local
+development. By convention, projects load configuration values from both:
+
+- a `.env.shared` file with common configuration values, and
+- a `.env` file with any developer-specific overrides (or empty)
+
+These files are excluded from source-control. The setup command initializes them, updating `.env.shared` from a
+private S3 location.
+
+Local ruby development depends on the [`foreman`](https://github.com/ddollar/foreman) utility and a `.foreman` file
+([e.g.](https://github.com/artsy/horizon/blob/master/.foreman)) to respect these files. Local node.js development
+depends on [node-foreman](https://github.com/strongloop/node-foreman) and the `--env` argument
```

Comment:
@joeyAghion this bit surprised me - I've never even heard of `node-foreman`. Is that because it's so baked-in I've never noticed it/because I've mostly worked on Artsy apps that are Ruby on Rails-based?

Reply:
It [is baked into](https://github.com/artsy/metaphysics/blob/245fd4e3740759b7077c438f6ba3f4ae5b771144/package.json#L19) commands like Metaphysics' `yarn dev`.

`dotenv` was the other utility we evaluated for this purpose, but we found that it doesn't support loading from multiple `.env*` files with correct precedence rules. (Having a shared source of common config and an isolated place for any individual developers' overrides was a goal of our work.)

Reply:
I'm also quite surprised we are using `node-foreman`. Personally, I would drop it in favor of `dotenv` to reduce the complexity added by `node-foreman` and reduce the risk that someone attempts to use it in production.

Reply:
I welcome help with this, actually. I wasn't able to get `dotenv` to work the same way that docker's `env_file` does with the 2 files. That's the basic need.